go-controller/pkg/node/controllers/egressip/egressip.go

```diff
@@ -0,0 +1,1130 @@
+			if !found {
+				existingNs = map[ktypes.NamespacedName]*podIPConfigList{}
+				existing.namespacesWithPodIPConfigs[updatedTargetNS] = existingNs
+			}
+			for updatedPodNamespacedName, updatedPod := range updatedTargetPod {
+				existingTargetPodConfig, found := existingNs[updatedPodNamespacedName]
+				if !found {
+					existingTargetPodConfig = newPodIPConfigList()
+					existingNs[updatedPodNamespacedName] = existingTargetPodConfig
+				}
+				// applyPodConfig will apply pod specific configuration - ip rules and iptables rules
+				err := c.applyPodConfig(existingTargetPodConfig, update)
+				if err != nil {
+					return fmt.Errorf("failed to apply pod %s/%s configuration for EgressIP %s IP %s: %v",
+						updatedPod.Namespace, updatedPod.Name, update.name, update.ip.String(), err)
+				}
+			}
+		}
+		// TODO(mk): only apply the follow when its new config or when it failed to apply
+		// Ok to repeat requests to route manager and link manager
+		if err := c.linkManager.AddAddress(*update.ip); err != nil {
+			return fmt.Errorf("failed to add address EgressIP %s IP %s to link manager: %v", update.name,
+				update.ip.String(), err)
+		}
+		existing.ip = update.ip
+		// route manager manages retry
+		c.routeManager.Add(*update.routeLink)
+		existing.routeLink = update.routeLink
+	}
+	return nil
+}
+
+func (c *Controller) applyPodConfig(existingConfig *podIPConfigList, updatedPolicy *config) error {
+	configToAdd := newPodIPConfigList()
+	for _, newConfig := range updatedPolicy.podInfos.elems {
+		if !existingConfig.hasWithoutError(newConfig) {
+			configToAdd.Insert(*newConfig)
+		}
+	}
+	for _, newConfig := range configToAdd.elems {
+		if err := c.ruleManager.Add(newConfig.ipRule); err != nil {
+			existingConfig.InsertOverwriteFailed(*newConfig)
+			return err
+		}
+		// v4
+		if newConfig.v6 {
+			if err := c.iptablesManager.EnsureRules(utiliptables.TableNAT, iptChainName, utiliptables.ProtocolIPv6, []iptables.RuleArg{newConfig.ipTableRule}); err != nil {
+				existingConfig.InsertOverwriteFailed(*newConfig)
+				return fmt.Errorf("unable to ensure iptables rules: %v", err)
+			}
+		} else {
+			if err := c.iptablesManager.EnsureRules(utiliptables.TableNAT, iptChainName, utiliptables.ProtocolIPv4, []iptables.RuleArg{newConfig.ipTableRule}); err != nil {
+				existingConfig.InsertOverwriteFailed(*newConfig)
+				return fmt.Errorf("failed to ensure rules (%+v) in chain %s: %v", newConfig.ipTableRule, iptChainName, err)
+			}
+		}
+		existingConfig.insertOverwrite(*newConfig)
+	}
+	return nil
+}
+
+func (c *Controller) getAllEIPs() ([]*eipv1.EgressIP, error) {
+	eips, err := c.eIPLister.List(labels.Everything())
+	if err != nil {
+		return nil, fmt.Errorf("failed to list EgressIPs: %v", err)
+	}
+	return eips, nil
+}
+
+// RepairNode generates whats expected and what is seen on the node and removes any stale configuration. This should be
+// called at Controller startup.
+func (c *Controller) RepairNode() error {
```

Comment:
Another general comment not directly related to the repair. If we were able to build up the `existing` cache of the controller from the information the repair function is obtaining, that would probably mean less netlink calls on startup. Not sure if otherwise this would still be a performance concern, guess it might not be if it is just specific to the startup.